src/promptflow-devkit/promptflow/_sdk/_utilities/chat_utils.py

```diff
@@ -10,12 +19,127 @@ def construct_flow_absolute_path(flow: str) -> str:
     if enable_internal_features:
         query_dict.update({"enable_internal_features": "true", **url_params})
     query_params = urlencode(query_dict)
     return urlunparse(("http", f"127.0.0.1:{port}", "/v1.0/ui/chat", "", query_params, ""))
+
+
+content_cache = defaultdict(dict)
+file_meta_cache = defaultdict(dict)
+
+
+def any_file_update(monitor_target, ignore=None):
+    global file_meta_cache
+    updated = False
+
+    for sub in monitor_target.rglob("**/*"):
+        if not sub.is_file():
+            continue
+        if ignore and any(sub.as_posix().startswith(i) for i in ignore):
+            continue
+        key, value = sub.as_posix(), sub.stat().st_mtime
+        if key not in file_meta_cache or file_meta_cache[key] != value:
+            file_meta_cache[key] = value
+            updated = True
+    return updated
+
+
+def any_content_update(target_file, keys: List[str]):
+    global content_cache
+    dot_key = "*".join(keys)
+    cached_content = content_cache[target_file].get(dot_key, None)
+    current_content = json.loads(target_file.read_text(encoding=DEFAULT_ENCODING))
+    for key in keys:
+        if current_content is not None:
+            current_content = current_content.get(key, None)
+    if cached_content != current_content:
+        content_cache[target_file][dot_key] = current_content
+        return True
+    return False
+
+
+def terminate_service(process):
+    if isinstance(process, multiprocessing.Process):
+        process.terminate()
+        process.join()
+    elif isinstance(process, subprocess.Popen):
+        process.terminate()
+        process.wait()
+    else:
+        raise ValueError(f"Invalid process type {type(process)}")
+
+
+def start_chat_ui_service_monitor(
+    flow, *, serve_app_port, pfs_port, url_params, init=None, enable_internal_features=False, skip_open_browser=False
+):
+
+    flow_directory, flow_path_file = resolve_flow_path(flow)
+    ux_input_path = flow_directory / PROMPT_FLOW_DIR_NAME / UX_INPUTS_JSON
+
+    # update init to ux_inputs.json
+    current_ux_inputs = (
+        json.loads(ux_input_path.read_text(encoding=DEFAULT_ENCODING)) if ux_input_path.is_file() else {}
+    )
+    if UX_INPUTS_INIT_KEY not in current_ux_inputs:
+        current_ux_inputs[UX_INPUTS_INIT_KEY] = {}
+    # save init with different key given there can be multiple prompty flow in one directory
+    current_ux_inputs[UX_INPUTS_INIT_KEY][flow_path_file] = init
+    ux_input_path.write_text(json.dumps(current_ux_inputs, indent=4, ensure_ascii=False), encoding="utf-8")
+
+    # show url for chat UI
+    url_params["serve_app_port"] = serve_app_port
+    chat_page_url = construct_chat_page_url(
+        str(flow_directory / flow_path_file),
+        pfs_port,
+        url_params=url_params,
+        enable_internal_features=enable_internal_features,
+    )
+    print(f"You can begin chat flow on {chat_page_url}")
+    if not skip_open_browser:
+        webbrowser.open(chat_page_url)
```

Comment:
need to ensure LPFS started?

Reply:
there is a `invoke_pf_svc` call in cli code to ensure this. pfs_port is got there.